execute.py

```diff
@@ -0,0 +1,759 @@
+
+        complex_force_matrix = complex_mm_force_matrix
+        return complex_force_matrix
+
+    def _create_sampler_states(self, complex_positions):
+        """
+        simple utility function to generate complex and ligand sampler states
+        """
+        from openmmtools.states import SamplerState
+        complex_sampler_state = SamplerState(positions=complex_positions,
+                                             box_vectors=self.complex_system.getDefaultPeriodicBoxVectors())
+        ligand_positions = unit.Quantity(np.zeros([len(self.ligand_subset_indices), 3]), unit=unit.nanometer)
+        ligand_positions[self.ligand_subset_indices, :] = complex_sampler_state.positions[self.ligand_subset_indices, :]
+        ligand_sampler_state = SamplerState(positions=ligand_positions,
+                                            box_vectors=self.ligand_system.getDefaultPeriodicBoxVectors())
+
+        return complex_sampler_state, ligand_sampler_state
+
+    def simulate_baoab(self,
+                       _lambda_index,
+                       n_steps=1,
+                       gamma=1. / unit.picoseconds,
+                       dt=1.0 * unit.femtoseconds,
+                       metropolize=False):
+        """
+        Simulate n_steps of BAOAB, accumulating heat
+        
+        args
+            n_steps : int, default 1
+                number of steps of dynamics to run
+            gamma : unit.quantity.Quantity (units compatible with 1.0/unit.picoseconds)
+                collision rate
+            dt : unit.quantity.Quantity (units compatible with units.femtoseconds)
+                timestep
+        """
+        Q = 0
+        W_shads = 0.
+        _complex_sampler_state, _ligand_sampler_state = SamplerState.from_context(
+            self.complex_context), SamplerState.from_context(self.ligand_context)
+        x_unit, v_unit, t_unit, m_unit = unit.nanometers, unit.nanometers / unit.femtoseconds, unit.femtoseconds, unit.daltons
+        x, v = _complex_sampler_state.positions.value_in_unit(
+            unit.nanometers), _complex_sampler_state.velocities.value_in_unit(unit.nanometers / unit.femtoseconds)
+        dt = dt.value_in_unit(t_unit)
+        gamma = gamma * t_unit
+        mass_matrix = self.mass_matrix.value_in_unit(m_unit)
+
+        def compute_kinetic_energy(v):
+            return 0.5 * np.sum(
+                np.matmul(np.square(v.T), mass_matrix.T)) * m_unit * v_unit ** 2 * self.complex_thermostate.beta
+
+        def sample_maxwell_boltzmann():
+            return np.sqrt((1. / self.complex_thermostate.beta).value_in_unit(m_unit * v_unit ** 2)) * (
+                    np.random.randn(mass_matrix.shape[1], 3) / np.sqrt(mass_matrix)[0, :][:, None])
+
+        def V(v):
+            force = self._compute_hybrid_forces(_lambda_index).value_in_unit(m_unit * x_unit / (t_unit ** 2))
+            dv = (dt / 2.) * (force / mass_matrix[0, :][:, None])
```

Comment:
Could you please annotate the intent of the reshape / slicing logic in this line?